Add documentation changes for disk-based k-NN (#8246)
opensearch-project/documentation-website: _search-plugins/knn/approximate-knn.md

```json
GET my-knn-index-1/_search
{
  "size": 2,
  "query": {
    "knn": {
      "target-field": {
        "vector": [2, 3, 5, 6],
        "k": 2,
        "rescore" : {
          "oversample_factor: 1.2
        }
      }
    }
  }
}
```

or
```json
GET my-knn-index-1/_search
{
  "size": 2,
  "query": {
    "knn": {
      "target-field": {
        "vector": [2, 3, 5, 6],
        "k": 2,
        "rescore" : true
      }
    }
  }
}
```

> The `oversample_factor` is a floating point number between 0.0 and 100.0. `oversample_factor*k` will always be greater than or equal to 100 and less than or equal to 10,000.

Reviewer: nit: Worth mentioning the defaults again here just incase someone is skimming through and directly jumps on to this section

:--- | :--- 
`faiss` | If `nprobes` is present in a query, it overrides the value provided when creating the index.

### Rescoring quantized results using full precision

Quantization can be used to significantly reduce the memory footprint of a k-NN index. For more information about quantization, see [k-NN vector quantization]({{site.url}}{{site.baseurl}}/search-plugins/knn/knn-vector-quantization). Because some precision of the vector representations is lost during quantization, the distances computed will be approximate. This causes the overall recall of the search to decrease. 

> To improve recall while maintaining the memory savings of quantization, you can use a two-phased search approach. In the first phase, `oversample_factor * k` results are retrieved from an index using quantized vectors and the scores are approximated. In the second phase, the full-precision vectors of those `oversample_factor * k` results are loaded into memory from disk, and scores are recomputed against the full-precision query vector. The results are then reduced to the top k.

Reviewer 2: ```suggestion
To improve recall while maintaining the memory savings of quantization, you can use a two-phase search approach. In the first phase, `oversample_factor * k` results are retrieved from an index using quantized vectors and the scores are approximated. In the second phase, the full-precision vectors of those `oversample_factor * k` results are loaded into memory from disk, and scores are recomputed against the full-precision query vector. The results are then reduced to the top k.
```

The default rescoring behavior is determined by the `mode` and `compression_level` of the backing k-NN vector field:

- For the `in_memory` mode, no rescoring is applied by default.

> - For the `on_disk` mode, default rescoring is based on the configured `compression_level`. Each `compression_level` provides a default `oversample_factor`, specified in the following table.

Reviewer 2: ```suggestion
- For `on_disk` mode, default rescoring is based on the configured `compression_level`. Each `compression_level` provides a default `oversample_factor`, specified in the following table.
```

| Compression level | Default rescore `oversample_factor` |
|:------------------|:----------------------------------|
| `32x` (default)   | 3.0                               |
| `16x`             | 2.0                               |
| `8x`              | 2.0                               |
| `4x`              | No default rescoring             |
| `2x`              | No default rescoring             |

> To explicitly apply rescoring, provide the `rescore` parameter in a query on a quantized index and specify the `oversample_factor`:

Reviewer 2: ```suggestion
To explicitly apply rescoring, provide the `rescore` parameter in a quantized index query and specify the `oversample_factor`:
```

```json
GET my-knn-index-1/_search
{
  "size": 2,
  "query": {
    "knn": {
      "target-field": {
        "vector": [2, 3, 5, 6],
        "k": 2,
        "rescore" : {
          "oversample_factor": 1.2
        }
      }
    }
  }
}
```
{% include copy-curl.html %}

Alternatively, set the `rescore` parameter to `true` to use a default `oversample_factor` of `1.0`:

```json
GET my-knn-index-1/_search
{
  "size": 2,
  "query": {
    "knn": {
      "target-field": {
        "vector": [2, 3, 5, 6],
        "k": 2,
        "rescore" : true
      }
    }
  }
}
```
{% include copy-curl.html %}

> The `oversample_factor` is a floating-point number between 1.0 and 100.0, inclusive. The number of results in the first pass is calculated as `oversample_factor * k` and is guaranteed to be between 100 and 10,000, inclusive. If the calculated number of results is less than 100, the number of results is set to 100. If the calculated number of results is greater than 10,000, the number of results is set to 10,000.

Reviewer 2: ```suggestion
The `oversample_factor` is a floating-point number between 1.0 and 100.0, inclusive. The number of results in the first pass is calculated as `oversample_factor * k` and is guaranteed to be between 100 and 10,000, inclusive. If the calculated number of results is smaller than 100, then the number of results is set to 100. If the calculated number of results is greater than 10,000, then the number of results is set to 10,000.
```
